[mgs] PendingMgsUpdate type to have a variant for the RoT

karencfv opened the pull request:
Implements the first step of updating the PendingMgsUpdate type to have a variant for the RoT. This lays out the foundation for the work in https://github.com/oxidecomputer/omicron/issues/7989

karencfv commented on nexus/mgs-updates/src/rot_updater.rs:
```diff
@@ -192,3 +197,29 @@ impl SpComponentUpdater for RotUpdater {
+pub struct ReconfiguratorRotUpdater;
+impl SpComponentUpdateHelper for ReconfiguratorRotUpdater {
+    /// Checks if the component is already updated or ready for update
+    fn precheck<'a>(
+        &'a self,
+        _log: &'a slog::Logger,
+        _mgs_clients: &'a mut MgsClients,
+        _update: &'a PendingMgsUpdate,
+    ) -> BoxFuture<'a, Result<PrecheckStatus, PrecheckError>> {
+        // TODO-K: fill in the precheck
+        todo!()
+    }
+
+    /// Attempts once to perform any post-update actions (e.g., reset the
+    /// device)
+    fn post_update<'a>(
+        &'a self,
+        _log: &'a slog::Logger,
+        _mgs_clients: &'a mut MgsClients,
+        _update: &'a PendingMgsUpdate,
+    ) -> BoxFuture<'a, Result<(), GatewayClientError>> {
+        // TODO-K: fill in the post_update
+        todo!()
+    }
```
These will be completed in a follow up PR (step two of https://github.com/oxidecomputer/omicron/issues/7989)

karencfv commented:
@jgallagher, @lzrd adding either of you as reviewers since Dave will be out for a bit :)

lzrd commented:
You are only covering RoT Hubris here, which is fine. I assume that RoT stage0 will be in later PRs.

lzrd commented:
There should be some provision for ordering of updates. Stage0, then RoT Hubris, then SP Hubris. Also the restriction that only one RoT stage0 activation is in progress in the entire rack at a time.
These constraints will be in later PRs but do you need to carry any of the supporting information in these structures at this time? (probably not).

jgallagher commented on nexus/types/src/deployment.rs:
```diff
@@ -1245,6 +1246,37 @@ pub enum PendingMgsUpdateDetails {
+    /// the RoT is being updated
+    Rot {
+        // implicit: component = ROT
+        // TODO-K: Is this accurate for the RoT as well?
+        // implicit: firmware slot id = 0 (always 0 for ROT)
```
It's not, but I think you've already addressed this in this structure: `active_slot` tells you which one is currently active, so the other one is the inactive slot that needs to be updated.

jgallagher commented on nexus/types/src/deployment.rs:
```diff
@@ -1245,6 +1246,37 @@ pub enum PendingMgsUpdateDetails {
+        /// expected contents of "B" slot
+        expected_slot_b_version: ExpectedVersion,
+        /// the slot of the currently running image
+        active_slot: RotSlot,
```
Should all of these fields have the prefix `expected_`?

jgallagher reviewed (commented):
> There should be some provision for ordering of updates. Stage0, then RoT Hubris, then SP Hubris. Also the restriction that only one RoT stage0 activation is in progress in the entire rack at a time.
These constraints will be in later PRs but do you need to carry any of the supporting information in these structures at this time? (probably not).

I think all of this is the responsibility of the planner; this work is about how to execute an update once the planner has decided it should happen.

karencfv commented on nexus/mgs-updates/src/driver.rs:
```diff
@@ -488,6 +504,21 @@ impl MgsUpdateDriver {
 }
 
+/// the component reconfigurator will update
+enum ReconfiguratorUpdater {
```
Ah yes, that's much nicer. Thanks!

karencfv commented:
Thanks for taking a look both! I think I've addressed all comments.

> You are only covering RoT Hubris here, which is fine. I assume that RoT stage0 will be in later PRs.

Yep! only RoT for now

jgallagher commented on dev-tools/reconfigurator-sp-updater/src/main.rs:
```diff
@@ -338,9 +338,10 @@ fn cmd_config(
                                             persistent_boot_preference {:?}
                                             pending_persistent_boot_preference {:?}
                                             transient_boot_preference {:?}",
```
Nit - probably want the `expected_` prefixes in the string labels too?

jgallagher commented on nexus/types/src/deployment.rs:
```diff
@@ -1319,19 +1318,22 @@ impl slog::KV for PendingMgsUpdateDetails {
                 )?;
                 serializer.emit_str(
                     Key::from("active_slot"),
```
Same nit here on the field labels

jgallagher reviewed (approved):
Thanks! Just a couple tiny renaming nits.
